rfcs/0014-abi-errors.md

```diff
@@ -0,0 +1,456 @@
+meaningful.
+
+It also pollutes the namespace with yet another reserved word.
+
+And ultimately it moves more complexity to the compiler rather than let the user define it.
+This is a shared problem with the error type annotations.
+
+# Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Despite the drawbacks of moving more special behavior the compiler, it has to
+shoulder this complexity in Sway since we do not have sufficiently powerful
+meta-programming features such that developers may be able to define their own
+language features of this caliber.
+
+Given the benefits of having standardized errors, these trade-offs seem
+acceptable, so long as we allow the solution to be extended in the future to
+be more user defined. The `Error` trait is the main vector of such future
+changes in this case.
+
+Letting the errors be populated directly in the bytecode would be a lot simpler,
+but it is not desirable because it extends the size of the binary for no gain.
+
+Moreover, the question of the location of error strings can be asked. Why
+put them in the ABI specification file? ABI files are soon to benefit from
+a registry that will allow users to easily fetch them, they already contain
+type metadata that can be extended, and they ultimately are our equivalent to
+headers, encompassing all the required information to call a contract and get
+a meaningful answer. Errors are part of making this answer meaningful and it is
+typical for error messages to live in headers.
+
+If we do not specify some kind of error standard, debugging any sort of complex
+Sway error will remain extremely difficult and supporting Sway contracts may be
+a nigh impossible task.
+
+# Prior art
+
+[prior-art]: #prior-art
+
+This RFC is somewhat inspired by Rust's error library ecosystem. We do not
+have enough metaprogramming facilities to allow users to define their own way
+of producing errors, but one can identify two kinds of Rust error handling
+libraries. Those that like `anyhow` care most about letting the user add some
+string based error context. And those like `thiserror` that allow one to build
+complex error types.
+
+We attempt here to cater to both crowds through `panic`'s dual modes.
+
+Moreover, Ethereum style embedding of errors in a `b256` isn't practical for us
+or generally desirable since the FuelVM uses 64 bit error codes and we aim to
+minimize the footprint of errors on deployed contracts.
+
+# Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+It's unclear how much of the existing special error codes produced by the
+standard library (such as `FAILED_ASSERT_SIGNAL`) should be migrated to use this
+mechanism or should remain special behavior.
+
+The extent of the SDK integration of this feature is also quite open ended. But
+implementation of this should give a good enough base to decide how much support
+we want and to extend it in the future.
+
+# Future possibilities
```

Comment:
One additional nice future possibility to mention explicitly. The proposed approach fits greatly in a possible language expansion for the recoverable errors, namely, having the propagating question mark operator `?` like in Rust.